File: zlib/bench/bench.py
"""Zlib benchmark in Python with gzip and compression.zstd (see README.md). The standard library has no brotli."""
import gzip
import time
from compression import zstd
from pathlib import Path

GZ = Path("out/payload.gz").read_bytes()
ZST = Path("out/payload.zst").read_bytes()
PLAIN = Path("out/plain.bin").read_bytes()


def chk(b: bytes) -> int:
    h = 0
    for x in b:
        h = (h * 31 + x) & 0xFFFFFFFF
    return h


t0 = time.perf_counter()
out = gzip.decompress(GZ)
ms = (time.perf_counter() - t0) * 1000
print(f"inflate\t{ms:.3f}\t{chk(out)}")

t0 = time.perf_counter()
packed = gzip.compress(PLAIN, compresslevel=6, mtime=0)
ms = (time.perf_counter() - t0) * 1000
print(f"deflate\t{ms:.3f}\t{chk(gzip.decompress(packed))}")
print(f"size\t0\t{len(packed)}")

t0 = time.perf_counter()
out = zstd.decompress(ZST)
ms = (time.perf_counter() - t0) * 1000
print(f"zstd\t{ms:.3f}\t{chk(out)}")

File: zlib/bench/bench.ts
// Run with node or bun, from this folder, after run.py has written out/payload.gz and out/plain.bin.
import { readFileSync } from "node:fs";
import { gunzipSync, gzipSync } from "node:zlib";

const gz = readFileSync("out/payload.gz");
const plain = readFileSync("out/plain.bin");

function chk(b: Buffer): number {
  let h = 0;
  for (let i = 0; i < b.length; i++) h = (Math.imul(h, 31) + b[i]!) >>> 0;
  return h;
}

let t0 = performance.now();
const out = gunzipSync(gz);
let ms = performance.now() - t0;
console.log(`inflate\t${ms.toFixed(3)}\t${chk(out)}`);

t0 = performance.now();
const packed = gzipSync(plain, { level: 6 });
ms = performance.now() - t0;
console.log(`deflate\t${ms.toFixed(3)}\t${chk(gunzipSync(packed))}`);
console.log(`size\t0\t${packed.length}`);
